[FSDP()][23/N] Refactor handle attr initialization

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88260 [FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`
* #88235 [FSDP()] Have `fully_shard()` abide by `@contract`!
* #88234 [FSDP()][Easy] Rename `_State` to `_FSDPState`
* #88233 [FSDP()] Rename to `fully_shard()` and move to `_composable/`
* #88232 [FSDP][Easy] Remove unneeded `TrainingState` transition
* #88123 [FSDP] Rename `unflat_param_name` -> `fqn` for consistency
* #88122 [FSDP] Simplify `_get_buffer_names()`
* #88121 [FSDP] Remove unneeded `torch.no_grad()` context when offloading to CPU
* #88120 [FSDP][Docs] Add note mentioning rate limiter for backward prefetch
* #88040 [FSDP()][27/N] Add forward hook registration
* #87941 [FSDP()][26/N] Move `_lazy_init()` into `_fsdp_root_pre_forward()`
* #87940 [FSDP()][25/N] Add `_post_forward_reshard()`
* #87939 [FSDP()][24/N] Refactor `_lazy_init()`
* **#87938 [FSDP()][23/N] Refactor handle attr initialization**
* #87935 [FSDP()][21/N] Refactor and fix `_cast_buffers()`
* #87934 [FSDP] Rename `dtype` to `buffer_name_to_dtype`
* #87933 [FSDP] Remove `device` arg from `_cast_buffers()`
* #87932 [FSDP()][20/N][Easy] Move functions in file
* #87931 [FSDP()][18/N] Refactor `pre_forward_unshard()`
* #87930 [FSDP()][17/N] Refactor `_fsdp_root_pre_forward()`
* #87929 [FSDP()][16/N] Refactor post-forward/pre-backward
* #87928 [FSDP()][15/N] Refactor `_init_streams()`
* #87927 [FSDP()][14/N] Refactor pre-forward/post-backward

**`_init_param_attributes()` -> `init_flat_param_attributes()`**
We move `_init_param_attributes()` to `FlatParamHandle.init_flat_param_attributes()` (as already marked as to-do during previous refactoring).

**`_reset_lazy_init()`**
We no longer delete `_local_shard` from each `FlatParameter` in `_reset_lazy_init()`.

**Analysis**
Thus, the two semantic differences are that we remove the initial `if hasattr(p, "_local_shard")` early return in `_init_param_attributes()` and the `delattr(p, "_local_shard")` in `_reset_lazy_init()`.

This is safe because
- If we never call `_reset_lazy_init()`, then `init_flat_param_attributes()` is only called once. There is no opportunity for an early return.
- If we call `_reset_lazy_init()`, then `init_flat_param_attributes()` will be called again in the next `_lazy_init()`. However, since we removed the early return, all of the attributes initialized in `init_flat_param_attributes()` simply get re-initialized and override any existing attributes.

Commit:
[FSDP()][23/N] Refactor handle attr initialization

[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -678,6 +682,73 @@ def shard_metadata(
+        For each tensor attribute on the ``FlatParameter``, see the unshard and
+        reshard methods in this class for the allocation and free pattern.
+        """
+        flat_param = self.flat_param
```
would I be correct that this is the same implementation except that "_local_shard" checks are removed?

awgu replied:
That is my bad. I mistakenly thought that the `_local_shard` checks are not important, but they are covered by unit tests. Let me include those back in.

Thanks for the close review!

awgu replied:
Update: Nevermind, the failing tests are because I reworded the error message below. The checks I removed are indeed able to be safely removed because of the invariants around `_lazy_init()` mentioned at https://github.com/pytorch/pytorch/pull/87937/files#r1009612359.

mrshenli reviewed (approved):
Unblock, assuming the only change is removing checks on `_local_shard`.

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][23/N] Refactor handle attr initialization"


**`_init_param_attributes() -> `init_flat_param_attributes()`**
We move `_init_param_attributes()` to `FlatParamHandle.init_flat_param_attributes()` (as already marked as to-do during previous refactoring).

**`_reset_lazy_init()`**
We no longer delete `_local_shard` from each `FlatParameter` in `_reset_lazy_init()`.

Thus, the two semantic difference are that we remove the initial `if hasattr(p, "_local_shard")` early return in `_init_param_attributes()` and the `delattr(p, "_local_shard")` in `_reset_lazy_init()`.

This is safe because
- If we never call `_reset_lazy_init()`, then `init_flat_param_attributes()` is only called once. There is no opportunity for an early return.
- If we call `_reset_lazy_init()`, then `init_flat_param_attributes()` will be called again in the next `_lazy_init()`. However, since we removed the early return, all of the attributes initialized in `init_flat_param_attributes()` simply get re-initialized and override any existing attributes.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
